### PR TITLE
Refactor generic parameter encoder functions

### PR DESCRIPTION
**Pull request by varkor**
Addresses https://github.com/rust-lang/rust/pull/58503#discussion_r257488950.

r? @eddyb

### CODE REVIEW

File: src/librustc_metadata/encoder.rs [hunk 1]
```diff
@@ -1302,10 +1303,11 @@ impl<'a, 'b: 'a, 'tcx: 'b> IsolatedEncoder<'a, 'b, 'tcx> {
...
-                                -> Entry<'tcx> {
-        debug!("IsolatedEncoder::encode_info_for_ty_param({:?})", def_id);
+    fn encode_info_for_generic_param(
+        &mut self,
+        (def_id, Untracked(encode_type)): (DefId, Untracked<bool>),
+    ) -> Entry<'tcx> {
+        debug!("IsolatedEncoder::encode_info_for_generic_param({:?})", def_id);
         let tcx = self.tcx;
         Entry {
             kind: EntryKind::Type,
```
Comment (eddyb) [inline src/librustc_metadata/encoder.rs]: This seems wrong, it should either vary with the kind of parameter, or be something like `EntryKind::GenericParam`.

File: src/librustc_metadata/encoder.rs [hunk 2]
```diff
@@ -1699,20 +1678,17 @@ impl<'a, 'b, 'tcx> IndexBuilder<'a, 'b, 'tcx> {
...
-                    self.record(def_id, encode_info, (def_id, has_default));
-                }
-                hir::GenericParamKind::Const { .. } => {
-                    let def_id = self.tcx.hir().local_def_id(param.id);
-                    let encode_info = IsolatedEncoder::encode_info_for_const_param;
-                    self.record(def_id, encode_info, def_id);
-                }
-            }
+            let encode_type = match param.kind {
+                GenericParamKind::Lifetime { .. } => continue,
```
Comment (eddyb) [inline src/librustc_metadata/encoder.rs]: Is this the right way to go about this? I guess `ty::Generics` are encoded as part of the generic item, and this part just handles extra information?

Does anything break if you only call `encode_info_for_generic_param` when `encode_type` is `true`?

File: src/librustc_metadata/encoder.rs (hunk 1 above)
Comment (varkor) [reply src/librustc_metadata/encoder.rs]: It's only recording any concrete types (e.g. defaults or a const's type), so I think `EntryKind::Type` is correct. But the name is misleading; I'll rename it to `encode_info_for_ty_in_generic_param` or something like that.

File: src/librustc_metadata/encoder.rs (hunk 2 above)
Comment (varkor) [reply src/librustc_metadata/encoder.rs]: > I guess `ty::Generics` are encoded as part of the generic item, and this part just handles extra information?

This part just handles type information.

> Does anything break if you only call `encode_info_for_generic_param` when `encode_type` is `true`?

Let's find out...

Comment (varkor) [reply src/librustc_metadata/encoder.rs]: > Does anything break
[Yes](https://github.com/rust-lang/rust/pull/58581#issuecomment-467643962).

File: src/librustc_metadata/encoder.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_metadata/encoder.rs]: A `static`/`const` item also has a type, but doesn't use `EntryKind::Type`.
IMO we should treat generic parameters as proper definitions.

File: src/librustc_metadata/encoder.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_metadata/encoder.rs]: But if you want to resolve this quickly, r=me and open an issue about generic parameters.

**Comment by varkor** [comment]
I've updated to take account of the different kinds.

**Review by eddyb — APPROVED** [review]

**Comment by Centril** [comment]
Speculatively assigning blame of failure in https://github.com/rust-lang/rust/pull/59457#issuecomment-477150165. @bors r-

**Comment by varkor** [comment]
I'll rebase on top of https://github.com/rust-lang/rust/pull/59466 and we'll see.

**Comment by cuviper** [comment]
You may be vindicated -- https://github.com/rust-lang/rust/pull/59466#issuecomment-477360976 failed the same way.

**Pull request merged**